### DISCUSSION TITLE
Welcome to perf-book Discussions!

### DISCUSSION THREAD

**Discussion by dendibakh** [dendibakh/perf-book#14] [Announcements]
## 👋 Welcome!
We’re using Discussions as a place to connect with contributors. Here you can ask questions, suggest improvements and share proposals for new content. Also, feel free to reach out to me via email: dendibakh@gmail.com.

**Comment by smthink** [comment]
Hi Denis, is this a public forum for discussions or a private one?

**Reply by dendibakh** [reply]
Hi Saurabh, it's public.

**Comment by kachini** [comment]
This book is related to CPU, but wondering if there any interest in persistent storage performance analysis topic i.e disks ? 
How to analyse the cases when system is not really busy.  What the system is waiting for? 
I can imagine topics like common mistakes, confusing metrics,  when should you run a benchmark and when should not..

**Reply by dendibakh** [reply]
Hi, thanks for the suggestions.
So far, disk performance is not on my radar.

> How to analyse the cases when system is not really busy. What the system is waiting for?

This is also not really on my radar. There is already an excellent book on this subject "Understanding Software Dynamics" by R. Sites.

> I can imagine topics like common mistakes, confusing metrics, when should you run a benchmark and when should not.

This is a good idea. I will think about it. Thanks!

**Comment by zbb136** [comment]
Hello, I am fortunate to have read your book and have benefited greatly. I have two questions about the perf tool that I would like to ask you:
1. In the example in your book, the parameter received by perf or toplev is an executable file (a.out), but how can I analyze its TMA using pid as a parameter?
2. The book focuses on TMA analysis in the process dimension. If I want to analyze TMA in the functional dimension, can I do it with the perf tool?

**Reply by dendibakh** [reply]
Hey @zbb136, sorry for the late reply. I saw it but then completely forgot to answer.

> 1. In the example in your book, the parameter received by perf or toplev is an executable file (a.out), but how can I analyze its TMA using pid as a parameter?

That's a good question, I'm surprised that there is no option to attach to some PID. You can ask this question here:
https://github.com/andikleen/pmu-tools/discussions
BTW, Linux perf has `-p, --pid` option for attaching.

> 2. The book focuses on TMA analysis in the process dimension. If I want to analyze TMA in the functional dimension, can I do it with the perf tool?

If you're looking for a per-function TMA breakdown, then VTune is the only choice. Alternatively, you can use marker APIs to collect needed perf events and calculate TMA metrics yourself. It's not easy but doable. Actually, check out [likwid](https://github.com/RRZE-HPC/likwid/wiki/TutorialMarkerC), it might already do what you want.

**Comment by mganguli** [comment]
Is there interest in adding performance benchmarking for HTTP apps like proxies? So CPU, memory, IO..?

**Reply by dendibakh** [reply]
Thank you for the suggestion, do you have particular workloads/benchmarks in mind? I don't discuss IO optimizations in the book, still thinking about it.
Let me ask you back a few questions just to understand it better...
What new information do you think it will bring? I.e. what readers will learn?

**Comment by plokhotnyuk** [comment]
If you have plans for a chapter about SWAR techniques then you can peek a lot of code samples from [jsoniter-scala](https://github.com/search?q=repo%3Aplokhotnyuk%2Fjsoniter-scala%20ByteArrayAccess&type=code) that  use them to parse/serialize numeric and date-time types from/to JSON representation.
